### ISSUE TITLE
"Crossbar has started event"

### ISSUE THREAD

**Issue by om26er** [crossbario/crossbar#1453]
I have an app that is based on crossbar, its a pluggable app that exposes functionalities of a Linux based system desktop over WAMP. This app also implements "network discovery" using [zeroconf](https://pypi.org/project/zeroconf/), currently it relies on the `node.pid` file to exist before broadcasting service name, this works sometimes but other times it sends the broadcast a little too early, causing a failed connection for clients.

We need a way to check if crossbar is fully running, something that's emitted after https://github.com/crossbario/crossbar/blob/master/crossbar/node/node.py#L482

A "hack" would be to just create an empty file inside `cbdir`, a more sophisticated solution might actually be to send some sort of event using twisted's reactor.

thoughts @meejah @oberstet ?

**Comment by om26er** [comment]
I am a noob when it comes to Twisted but I proposed a minor change that could address this case.

**Comment by oddjobz** [comment]
Hi, not sure if I'm totally understanding, but could you get away with something like this;

**Comment by oddjobz** [comment]
Hi, not sure if I'm totally understanding, but could you get away with something like this; adding this to config.json will enable a unix domain socket;
```xml
            "transports": [
                {
                    "type": "rawsocket",
                    "endpoint": {
                        "type": "unix",
                        "path": "/tmp/uds_watch.sock"
                    }
                },
```
With a twisted client;
```python
#!/usr/bin/env python3

from twisted.internet import protocol, reactor


class Monitor(protocol.Protocol):
    
    def connectionMade(self):
        print('Call "after startup" event trigger here')

    def connectionLost(self, reason):
        print('Call "after shutdown" event trigger here')


class MonitorFactory(protocol.ReconnectingClientFactory):

    protocol = Monitor
    maxDelay = 1
    initialDelay = 1

    def clientConnectionLost(self, connector, reason):
        print('Crossbar has gone down!')
        protocol.ReconnectingClientFactory.clientConnectionLost(self, connector, reason)

    def clientConnectionFailed(self, connector, reason):
        print('Waiting for Crossbar to come up ...')
        protocol.ReconnectingClientFactory.clientConnectionFailed(self, connector, reason)


if __name__ == '__main__':
    reactor.connectUNIX('/tmp/uds_watch.sock', MonitorFactory())
    reactor.run()
```

**Comment by meejah** [comment]
If your "service discovery" thing was a `Component`  you could just wait until it connects successfully, and then broadcast the zeroconf stuff (e.g. in the `@on_join` handler) since the use-case is "network things are listening".

Another option might be to use the systemd notifier stuff (that already exists). (This is basically like the inverse @oddjobz suggestion, because systemd passes a unix socket path in `NOTIFY_SOCKET` to which crossbar will connect and send `READY=1` -- actually uses the "sdnotify" package).

**Comment by om26er** [comment]
> If your "service discovery" thing was a `Component` you could just wait until it connects successfully, and then broadcast the zeroconf stuff (e.g. in the `@on_join` handler) since the use-case is "network things are listening".

That could actually work, I just hope crossbar's "in-router" components are loaded sequentially so that my "service discovery" thing gets connected last.

**Comment by meejah** [comment]
There is no order imposed, AFAIK. But I think the "best, crossbar-wide" solution here is an "application_ready" event which has a testament/event-memory of 1 -- then in `on_join` you listen for this and ask for any remembered event, meaning that if its already fired you'll get the publish right away.

It would be good to have a general pattern here, as it's definitely a common problem (not just for WAMP/crossbar) to answer "is my distributed system ready to go?". This is similar to the Twisted solution you've proposed, except that will only work for in-router things within the same event-loop -- a WAMP-based solution would be better.

Instead of "order", I've thought it might be good to be able to indicate dependencies in components. This would most logically be a CrossbarFX feature, though: the monitoring-thing can thus be asked *what* is ready to go, exactly (including "everything") and individual components would probably call some method in the monitoring infrastructure to indicate "ready" (presumably tied to their connection, i.e. they become "not ready" when their session goes away).

**Comment by oddjobz** [comment]
Mmm, I think because you were talking about node.pid and creating files, I sort of assumed you were trying to avoid making WAMP connection to Crossbar .. on_join would be a more natural solution. On CFC you 'could' call get_realms/get_realm to check the status of each realm / wait for started .. I don't know if this exists in OS Crossbar or if there is an equivalent?

**Comment by meejah** [comment]
You can call all the "WAMP Meta API" methods/events in OS crossbar. But there's nothing for realms, AFAIK. In any case, that's sort of a different feature than "is stuff ready" -- although for some use-cases perhaps "the correct realm is running" might be sufficient?

Part of the problem here, of course, is that "ready" is application-dependent. Perhaps some components are "optional" or not always present (e.g. debug or monitoring things) and there could be dependencies between components (e.g. "component A" might need to start after "component B"). There's also "stuff that Crossbar / WAMP can't even see" like maybe "component A" starts up and makes a WAMP connection but hasn't yet successfully connected to the database (so it has a WAMP session, but isn't "ready" yet).

There *IS* a notion of "ready" of sorts built into the `Component` lifecycle: these call `ready` notifications/callbacks when: a session is connected; `on_join` has completed (including any async work) **and** all `on('join')` listeners have completed (including any async work). Perhaps this could form the core of a more-cohesive notion of "ready" for the entire application..?

**Comment by oddjobz** [comment]
+1 :: concept of ready would be very useful IMO, if it can be nailed down .. :) .. in this case, I was working on the basis that the problem involved failed client connections .. so once all realms are started, hopefully no client connections would fail .. (?) .. that said, if all client connections connect to one realm and you use an on_join for that realm, that in itself should be enough .. (?)

**Comment by oberstet** [comment]
fwiw, here are some relevant code locations:

* https://github.com/crossbario/crossbar/blob/master/crossbar/worker/router.py#L261
* https://github.com/crossbario/crossbar/blob/master/crossbar/worker/router.py#L282
* https://github.com/crossbario/crossbar/blob/master/crossbar/router/service.py#L144

 `onready` (which is waited on when a realm is started before returning) and a WAMP event .. not sure if that is already providing the semantics @oddjobz is looking for?

**Comment by oberstet** [comment]
I just realized: are you looking for a way to hook _app_ code into the events above? because that won't work out of the box, as the management events are published on the management realm running on the node controller WAMP router - _not_ any app router started.

means: this needs more thought/design ..

**Comment by meejah** [comment]
Yeah, I was using the `Component` events mostly as an example. I think this *is* a common need for many applications/reasons but also the semantics are very likely application-specific, too. Simply "connected to a realm" is only going to be sufficient for some use-cases. For others, it'll be after some amount of work is done (e.g. "after `on_join` has run" can probably capture these sorts of semantics, like connecting to a database or similar). There will also be dependency-type semantics to capture (e.g. "I can only run once these WAMP APIs are available" .. which might be captured by "I can only run once these Components are running")...

**Comment by oberstet** [comment]
ok, I get it. yes, not only useful, but probably often required, but yeah, the desired semantics isn't straight forward - or at least there are different scenarios as you hint at.

fwiw, we currently have the `bridge_meta_api` configuration option, which exposes the WAMP meta API on a regular app router (realm) on the node controller router (the management realm there).

eg a session joining an app realm on a router worker will fire `wamp.session.on_join` (if I recall correctly the URI), but that event will be on that app realm only.

setting above option will additionally publish that event on the management realm (via the controller session from the router worker to the management router and via the stdio based WAMP transport between the 2 processes).

the use case here requires the other direction: when a management event is published on the node controller router / management realm, forward that over the controller WAMP transport(s) to the app router worker(s).

implementation wise this is doable in a couple of LOC - main thing is to sufficiently define what we acutally want (and how to configure that)

**Comment by meejah** [comment]
I think from a user's perspective we want a hook-able API which captures the following kinds of scenarios (we might not provide a complete solution here, out-of-the-box, e.g. maybe we're just providing the right hooks and an example pattern of how to use them in your application):

 - a single component is "ready" or not
 - components can express dependencies to other APIs/components
 - ...?

Assuming full access to the existing WAMP Meta API, could we write a "co-ordination" component that captures the above semantics?

It might make sense to avoid tying all this to "components" and concentrate on "WAMP URIs" instead? Ideally "a component" handles some cohesive set of APIs, but ... e.g. frame the above as:

 - particular APIs (registered WAMP URIs) are available (or not)
 - particular APIs depend on other APIs being available (before they can start and/or function)

So, let's see. For the first point:

 - we can write an in-router component that listens for ``wamp.registration.on_register`` (etc) and determine if a given API-collection is on-line (or not).
 - this same component could export its own API (e.g. publish a "ready" event with persistence or provide a "is_ready"-style function)

For the second point (dependent APIs), a pattern could be followed so that upon startup (and *before* registering any of its APIs) a component with dependencies would wait for the above component to signal that the correct APIs are ready -- either waiting for the publish or calling some kind of "is_ready()" or "await_ready()" or similar API).

**Comment by oberstet** [comment]
see also https://github.com/crossbario/crossbar/issues/1338
